### ISSUE TITLE
Improve propagation of gossip about new nodes

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#2862]
When new node is added, it gossips with the seed nodes to get the initial cluster topology. It does so using a shadow syn message, which includes no states. As a response, seeds will send back information about all other nodes, but will learn nothing about the new node. They can learn about it from the regular SYN message sent from a node which already knows about the new node.

It would be more robust if the seeds learnt about the new node quickly, so that nodes added later will immediately see it. This matters when growing a large cluster. 

Another improvement would be to allow nodes to learn about new nodes from SYN responses. Currently responses include only information about nodes which were in the SYN digest, so they can only be pushed by a SYN from a node which already knows about the new node. Initially such nodes are in minority.

Refs #2855

**Comment by elcallio** [comment]
While I most certainly don't have a clue about gossip stuff, would it sort of solve your "deliver info on myself soon" problem if a booting node on recieving the ACK for shadow round SYN, instead of doing nothing, instead _did_ send an ACK2 with just ifself in the ep state?

**Comment by tgrabiec** [comment]
@elcallio It could achieve this, but we should be careful not to gossip our state prematurely (whatever that means) so that it doesn't confuse other nodes. I can see that a shadow round is performed before the actual gossiping is started. Maybe we should just make the gossiper prefer seeds once gossiping is started, utilizing the existing mechanism for preferring newly added nodes as gossip targets (`_live_endpoints_just_added`)

**Comment by tgrabiec** [comment]
The name "shadow" suggests that we intentionally don't advertise ourselves until we successfully bootstrap.

**Comment by elcallio** [comment]
Not sure that "shadow" really "suggests" anything, but sure... However, preferring seed node, at least initially, already pretty much happens, since the initial round(s) in gossiper:run will  have empty live endpoint sets, and thus will instead send a syn to all known seeds (gossiper.cc:665). 
After this has any result, it seems `_live_endpoints_just_added` will be seed (+ maybe other) next round also. In any case, the end result should be that, assuming we tell about ourselves in ack2, a starting node should quite quickly present itself to at least the seed. 
So I'm then a little unsure exactly what you envision changed.

**Comment by tgrabiec** [comment]
2017-11-15 15:27 GMT+01:00 Calle Wilund <notifications@github.com>:

> Not sure that "shadow" really "suggests" anything, but sure... However,
> preferring seed node, at least initially, already pretty much happens,
> since the initial round(s) in gossiper:run will have empty live endpoint
> sets, and thus will instead send a syn to all known seeds (gossiper.cc:665).
>

Won't we learn about all live endpoint sets from the shadow round?

**Comment by elcallio** [comment]
One would think we did, but in my (naive) debug testing, we did not. I'll re-test, because it sounds weird indeed when I think about it.

**Comment by elcallio** [comment]
Nope, sorry, I was right. gossiper.cc:1316. In shadow run we never add returned nodes to live endpoint set. Thus my earlier statement about us initially only sending syn to known seeds is correct. 
Whether or not this is as wanted/intended I'll leave unsaid.

**Comment by tgrabiec** [comment]
Ok, so it looks like the first part is already implemented.

**Comment by elcallio** [comment]
For the second point, while we get a full ack response to the initial shadow run, we clear the endpoint states before continuing bootstrap. Thus once we actually start gossiping, we only as about our selves to the seed. It would seem it is not until we actually receive ack2:s (i.e. reverse gossip) we actually start knowing about other endpoints. 
Seems a simple fix would be to just check "if we are only querying about ourselves, send an empty SYN instead" (shadow, but not). We then immediately get all info from the seed. Quick modification locally seems to work ok, but there might be some caveats?

**Comment by tgrabiec** [comment]
I don't understand. What does it mean to "ask about our selves" / "query about ourselves" above?

**Comment by elcallio** [comment]
Since we clear the endpoint states on bootstrap, after shadow round, once we actually get to gossiper::run(), the map contains only ourselves. Thus the digests in our SYN sent is only for our own node. And we repeatedly ask the seed node for info about ourself, getting very little in return. After a while, we will get a SYN ourself, send an ACK with also very little in, but at least get an ACK2 with info on other nodes. Now we finally start asking about these in run. 
(All this from log/debugger inspection on what is happing on a bootstraping a non-seed node in my tiny 3-node cluster - ymmd). 
I seems to me that since we can easily see in run() that we are asking about only ourselves, we can just turn the query into a "ask-about-all", i.e. empty SYN. Seems to get us info much quicker.

**Comment by tgrabiec** [comment]
We shouldn't make every SYN empty, because it means we will sent over network all info in every gossip round. I think we should still send digests about all we know, but if the other side knows about nodes which we don't know, it should also include that info in ACK response, considering it as a digest mismatch.

**Comment by elcallio** [comment]
I am not suggesting making every SYN empty. Just the initial ones where we only as for ourselves (one or two depending on timing). All others should continue as usual.

**Comment by tgrabiec** [comment]
If node does that only in the initial SYN, then it won't learn any faster about new nodes which are added later.
